homeassistant/components/emulated_hue/upnp.py

```diff
@@ -53,105 +55,94 @@ def get(self, request):
         return web.Response(text=resp_text, content_type="text/xml")
 
 
-class UPNPResponderThread(threading.Thread):
-    """Handle responding to UPNP/SSDP discovery requests."""
+@core.callback
+def create_upnp_datagram_endpoint(
+    host_ip_addr, upnp_bind_multicast, advertise_ip, advertise_port,
+):
+    """Create the UPNP socket and protocol."""
 
-    _interrupted = False
+    # Listen for UDP port 1900 packets sent to SSDP multicast address
+    ssdp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    ssdp_socket.setblocking(False)
 
-    def __init__(
-        self,
-        host_ip_addr,
-        listen_port,
-        upnp_bind_multicast,
-        advertise_ip,
-        advertise_port,
-    ):
-        """Initialize the class."""
-        threading.Thread.__init__(self)
+    # Required for receiving multicast
+    ssdp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+    ssdp_socket.setsockopt(
+        socket.SOL_IP, socket.IP_MULTICAST_IF, socket.inet_aton(host_ip_addr)
+    )
+
+    ssdp_socket.setsockopt(
+        socket.SOL_IP,
+        socket.IP_ADD_MEMBERSHIP,
+        socket.inet_aton(BROADCAST_ADDR) + socket.inet_aton(host_ip_addr),
+    )
+
+    ssdp_socket.bind(("" if upnp_bind_multicast else host_ip_addr, BROADCAST_PORT))
+
+    loop = asyncio.get_event_loop()
 
-        self.host_ip_addr = host_ip_addr
-        self.listen_port = listen_port
-        self.upnp_bind_multicast = upnp_bind_multicast
+    return loop.create_datagram_endpoint(
```

Comment:
This code shouldn't be in the integration at all, but in a separate library.

Reply:
Maybe home-assistant-libs?

Reply:
Yes, I've gotten ok for that. Is that something you want to do? The upnp socket should be easy to extract. The rest of the hue api is probably harder to move, so I'd be happy if we can move the upnp code at least.

Reply:
I'm happy to do it in a few weeks once I've dug myself out of what I'm currently working on and `WTH` month is over.

If we can get the repo created in the mean time, that would be great.

Reply:
https://github.com/home-assistant-libs/emulated-hue

Let me know if you think of a better name.